io/plugins/eu.esdihumboldt.hale.io.geopackage/src/eu/esdihumboldt/hale/io/geopackage/GeopackageInstanceWriter.java

```diff
@@ -0,0 +1,570 @@
+import eu.esdihumboldt.hale.common.schema.geometry.GeometryProperty;
+import eu.esdihumboldt.hale.common.schema.model.DefinitionUtil;
+import eu.esdihumboldt.hale.common.schema.model.PropertyDefinition;
+import eu.esdihumboldt.hale.common.schema.model.TypeDefinition;
+import eu.esdihumboldt.hale.common.schema.model.constraint.type.Binding;
+import eu.esdihumboldt.hale.common.schema.model.constraint.type.GeometryMetadata;
+import eu.esdihumboldt.hale.common.schema.model.constraint.type.GeometryType;
+import eu.esdihumboldt.hale.common.schema.model.constraint.type.PrimaryKey;
+import mil.nga.geopackage.BoundingBox;
+import mil.nga.geopackage.GeoPackage;
+import mil.nga.geopackage.attributes.AttributesColumn;
+import mil.nga.geopackage.attributes.AttributesDao;
+import mil.nga.geopackage.attributes.AttributesRow;
+import mil.nga.geopackage.core.srs.SpatialReferenceSystem;
+import mil.nga.geopackage.core.srs.SpatialReferenceSystemDao;
+import mil.nga.geopackage.db.GeoPackageDataType;
+import mil.nga.geopackage.features.columns.GeometryColumns;
+import mil.nga.geopackage.features.user.FeatureColumn;
+import mil.nga.geopackage.features.user.FeatureDao;
+import mil.nga.geopackage.features.user.FeatureRow;
+import mil.nga.geopackage.geom.GeoPackageGeometryData;
+import mil.nga.geopackage.manager.GeoPackageManager;
+import mil.nga.geopackage.schema.TableColumnKey;
+import mil.nga.geopackage.user.UserColumn;
+import mil.nga.geopackage.user.UserCoreRow;
+import mil.nga.sf.util.ByteReader;
+import mil.nga.sf.wkb.GeometryReader;
+
+/**
+ * Writes features to a GeoPackage file.
+ * 
+ * @author Simon Templer
+ */
+public class GeopackageInstanceWriter extends AbstractInstanceWriter {
+
+	@Override
+	public boolean isPassthrough() {
+		return isWriteUnordered();
+	}
+
+	/**
+	 * @return if instances are written unordered
+	 */
+	protected boolean isWriteUnordered() {
+		// FIXME support properly handling foreign key constraints (similar to
+		// for JDBC)?
+		// XXX for now just a simple unordered write
+		return true;
+	}
+
+	@Override
+	public boolean isCancelable() {
+		return false;
+	}
+
+	@Override
+	protected IOReport execute(ProgressIndicator progress, IOReporter reporter)
+			throws IOProviderConfigurationException, IOException {
+		InstanceCollection instances = getInstances();
+
+		GeoPackage geoPackage = null;
+		try {
+			URI loc = getTarget().getLocation();
+			File file;
+			try {
+				file = new File(loc);
+			} catch (Exception e) {
+				throw new IllegalArgumentException("Only files are supported as data source", e);
+			}
+
+			if (file.exists() && file.length() == 0L) {
+				// convenience for overwriting empty existing file
+				file.delete();
+			}
+			if (!file.exists()) {
+				GeoPackageManager.create(file);
+			}
+			// open existing file
+			geoPackage = GeoPackageManager.open(file, true);
+
+//			geoPackage.beginTransaction(); //XXX not sure how to use this -> results in a SQL exception later on (SQLITE_BUSY)
```

Comment:
There is a test for the transactions: https://github.com/ngageoint/geopackage-java/blob/master/src/test/java/mil/nga/geopackage/test/db/TransactionTest.java

Reply:
The key is probably disabling autocommit, I'll test that later

Reply:
I played around a bit, but it seems it's not so easy. Could be that the problem is that I want to both create tables and insert data in the transaction (a different approach would require some refactoring). As writing w/o transaction works well, I would not see that as a priority right now.